// ngAdapter/build/ngAdapter.js
"use strict";
var upgrade_1 = require('angular2/upgrade');
var downgrade_1 = require('./downgrade');
var upgrade_2 = require('./upgrade');
var helper_1 = require('./helper');
var scopeevents_1 = require('./scopeevents');
var ngAdapter = (function () {
    function ngAdapter(module) {
        this.module = module;
        this.addedProviders = {};
        this.upgradedProviders = [];
        this.upgradeAdapter = new upgrade_1.UpgradeAdapter();
        this.downgrade = new downgrade_1.Downgrade(this.upgradeAdapter, module);
        this.upgrade = new upgrade_2.Upgrade(this.upgradeAdapter, module, this.addedProviders, this.upgradedProviders);
        this.addProvider(scopeevents_1.ScopeEvents);
    }
    ngAdapter.prototype.bootstrap = function (element, app) {
        return this.upgradeAdapter.bootstrap(element, app);
    };
    ngAdapter.prototype.downgradeNg2Component = function (cmp) {
        return this.upgradeAdapter.downgradeNg2Component(cmp);
    };
    ngAdapter.prototype.downgradeNg2Directive = function (directive) {
        return this.downgrade.directive(directive);
    };
    ngAdapter.prototype.downgradeNg2Provider = function (provider) {
        return this.upgradeAdapter.downgradeNg2Provider(provider);
    };
    ngAdapter.prototype.addProvider = function (provider) {
        this.addedProviders[helper_1.getFunctionName(provider)] = provider;
        return this.upgradeAdapter.addProvider(provider);
    };
    ngAdapter.prototype.upgradeNg1Component = function (directive) {
        return this.upgradeAdapter.upgradeNg1Component(directive);
    };
    ngAdapter.prototype.upgradeNg1Directive = function (directive) {
        //let upgrade = new Upgrade(this.upgradeAdapter, this.module, this.addedProviders,
        //  this.upgradedProviders);
        return this.upgrade.upgradeNg1Directive(directive);
    };
    ngAdapter.prototype.upgradeNg1Provider = function (provider) {
        this.upgradedProviders.push(provider);
        return this.upgradeAdapter.upgradeNg1Provider(provider);
    };
    return ngAdapter;
}());
exports.ngAdapter = ngAdapter;
